### PR TITLE
Model export update.

### PR DESCRIPTION
**Pull request by tomasatdatabricks**
1. mlflow is installed from pip inside the container, unless MLFLOW_DEV environment variable is set. This applies to both azureml and sagemaker containers.
2. the sagemaker init script is translated to python. 
3. refactored mlflow sagemaker code. 
   - python api moved to sagemaker/__init__. This is to provide python api (cli 
      targets can not be easily called from python) usable e.g. from notebooks.
4. translated some of sh script to build and push image to ecr to python.  
5. added test for mlflow sagemaker run-local.

### CODE REVIEW

File: mlflow/pyfunc/scoring_server.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def transformation():  # noqa
             data = pd.read_json(s, orient="records")
         else:
             return flask.Response(
-                response='This predictor only supports CSV of josn  data, got %s' % str(
+                response='This predictor only supports CSV of json  data, got %s' % str(
```
Comment (mateiz) [inline mlflow/pyfunc/scoring_server.py]: "CSV of json" -> "CSV or JSON"

File: mlflow/sagemaker/__init__.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+            tar.add(os.path.join(source_dir, f), arcname=f)
+
+
+def _upload_s3(local_model_path, bucket, prefix):
+    """
+    Upload dir to S3 as .tar.gz.
+    :param local_model_path: local path to a dir.
+    :param bucket: S3 bucket where to store the data.
+    :param prefix: path within the bucket.
+    :return:
```
Comment (mateiz) [inline mlflow/sagemaker/__init__.py]: Do we need `:return:` if it won't return anything? same below

File: mlflow/sagemaker/__init__.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+
+def _deploy(role, image, app_name, model_s3_path, run_id):
+    """
+    Deploy model on sagemaker.
+    :param role:
+    :param image:
+    :param app_name:
+    :param model_s3_path:
+    :param run_id:
```
Comment (mateiz) [inline mlflow/sagemaker/__init__.py]: Missing docs

File: mlflow/sagemaker/container/__init__.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+from mlflow import pyfunc
+from mlflow.models import Model
+
+
+cpu_count = multiprocessing.cpu_count()
+model_server_timeout = os.environ.get('MODEL_SERVER_TIMEOUT', 60)
+model_server_workers = int(os.environ.get('MODEL_SERVER_WORKERS', cpu_count))
+
+
+def _scoring_server_init(cmd):
```
Comment (mateiz) [inline mlflow/sagemaker/container/__init__.py]: What can "cmd" be here?

File: mlflow/sagemaker/container/__init__.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
+from __future__ import print_function
```
Comment (mateiz) [inline mlflow/sagemaker/container/__init__.py]: This file needs a doc comment saying when and where this runs

File: .travis.yml [hunk 6]
```diff
@@ -10,11 +14,13 @@ install:
   # (The conda installation steps below are taken from http://conda.pydata.org/docs/travis.html)
   # We do this conditionally because it saves us some downloading if the
   # version is the same.
+  - find .
```
Comment (mateiz) [inline .travis.yml]: Do we need this or is it just for debugging?

File: mlflow/azureml/__init__.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+from mlflow.models import Model
+from mlflow.tracking import _get_model_log_dir
+from mlflow.utils.file_utils import TempDir
+from mlflow.version import VERSION as mlflow_version
+
+
+def deploy(app_name, model_path, run_id, mlflow_home):
+    """Deploy MLflow model to Azure ML.
+
+    This command will deploy MLflow model to Azure ML.
```
Comment (mateiz) [inline mlflow/azureml/__init__.py]: Do we need this line here? The first line said the same thing

File: mlflow/azureml/__init__.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+import mlflow
+from mlflow import pyfunc
+from mlflow.models import Model
+from mlflow.tracking import _get_model_log_dir
+from mlflow.utils.file_utils import TempDir
+from mlflow.version import VERSION as mlflow_version
+
+
+def deploy(app_name, model_path, run_id, mlflow_home):
+    """Deploy MLflow model to Azure ML.
```
Comment (mateiz) [inline mlflow/azureml/__init__.py]: Nit: I've preferred to keep the `"""` on its own line to align the text better in other docs

File: mlflow/azureml/__init__.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    :param app_name: Name of the deployed application
+    :param model_path: Local or mlflow-run-relative path to the model to be exported
+    :param run_id: If provided, run_id is used to retrieve the model logged with mlflow.
+    """
+    if run_id:
+        model_path = _get_model_log_dir(model_path, run_id)
+    model_path = os.path.abspath(model_path)
+    with TempDir(chdr=True, remove_on_exit=True):
+        exec_str = _export(app_name, model_path, mlflow_home=mlflow_home)
+        print("executing", '"{}"'.format(exec_str))
```
Comment (mateiz) [inline mlflow/azureml/__init__.py]: use `eprint` to print to stderr instead

File: .travis.yml (hunk 6 above)
Comment (tomasatdatabricks) [reply .travis.yml]: yes good catch, I'll remove it

File: mlflow/azureml/__init__.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    with open(score_py, "w") as f:
+        f.write(SCORE_SRC)
+
+    deps = ""
+    mlflow_dep = "mlflow=={}".format(mlflow_version)
+
+    if mlflow_home:
+        print("***")
+        print("MLFLOW_HOME =", mlflow_home)
+        print("***")
```
Comment (mateiz) [inline mlflow/azureml/__init__.py]: Same comment about `eprint`

File: mlflow/azureml/__init__.py (hunk 7 above)
Comment (tomasatdatabricks) [reply mlflow/azureml/__init__.py]: True, I'll remove it.

File: mlflow/utils/file_utils.py [hunk 11]
```diff
@@ -228,3 +229,37 @@ def write_to(filename, data):
 def append_to(filename, data):
     with open(filename, "a") as handle:
         handle.write(data)
+
+
+def _copy_project(src_path, dst_path=""):
```
Comment (mateiz) [inline mlflow/utils/file_utils.py]: You should document that this looks at a docker ignore file if one exists. Should it also do gitignore?

File: tests/mlflow/sagemaker/test_model_export.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+                path_to_remove = tmp._path
+                # NOTE: Changed dir to temp dir and use relative paths to get around the way temp
+                # dirs are handled in python.
+                model_pkl = tmp.path("model.pkl")
+                with open(model_pkl, "wb") as f:
+                    pickle.dump(self._linear_lr, f)
+                input_path = tmp.path("input_model")
+                conda_env = "conda.env"
+                import sys
+                python_version = sys.version.split("|")[0].strip()
```
Comment (mateiz) [inline tests/mlflow/sagemaker/test_model_export.py]: It's better to use sys.version to pull out the components; for example Mac OS's Python doesn't have any `|` in it

**Review by mateiz — CHANGES_REQUESTED** [review]

**Comment by mateiz** [comment]
FYI it looks like this doesn't merge anymore due to a small doc change I merged. Would you mind fixing that? Otherwise I think it's OK.

**Review by mateiz — APPROVED** [review]

**Pull request merged**